### DISCUSSION TITLE
Allow passing in my own node id instead of generating one internally

### DISCUSSION THREAD

**Discussion by mochar** [cachapa/crdt#13] [General]
I want to have access to the device-specific node id before the database is created, which is not possible as the node id is generated on db init. Therefore I think a solution could be generating the node id myself (with your `generateNodeId()` function) and passing that to the constructor/init function.

**Comment by mochar** [comment]
After reading the source code I'm a bit confused. The package seems to use the node ID of last modified HLC as the client's node ID. Is that right? I thought each client would have its own unique identifier?

**Reply by cachapa** [reply]
The `modified` timestamp is local to each node, and always uses the local node id since it represents the canonical timestamp at the time of writing.

Not to be confused with the `hlc` timestamp which represents the time at which the record was created / updated - that should be equal to `modified` for locally created but different for remote ones.

**Reply by mochar** [reply]
Thanks so much for the excellent explanation! I think a short summary in the readme would be nice. Especially if written in a way that doesn't require any prerequisite knowledge coming in. I don't mind writing this for you once I grasp the entire concept a bit better.

**Comment by cachapa** [comment]
I used to have this option in the past but ran against problems while using it myself enough times that I decided it would be best to not allow that as having duplicate node ids in a network will break the CRDT in sometimes subtle ways.

Can you give a bit more details on what you want to accomplish with this? I'm wondering if there's another way to do the same, or if your use case justifies allowing for that option.

**Reply by mochar** [reply]
Ah OK fair enough! 

My use case: I was messing around with a system where a folder with one sqlite db file per client is synced around to each client's device. Then, when a client notices that one of the other client's db file has changed, it would merge the changes to its own db. I thought it would be very clean to have the filename of the db be the node id. That means that I would need the node id before creating the db, which is currently not possible.

The justification for syncing all the db files on each client's device is because I don't want a server to act as a mediator for syncing db state across devices. I also don't want to implement a system where the clients can communicate with each other directly. I think this solution is fine because for my app, the user has to already have a system in place where files are being synced across devices, so the idea was to just add the dbs as another set of files you have to sync.

I changed the setup a bit now since all db files contain mostly redundant information, so it is a bit of a waste. Instead I have one sqlite file that gets synced, that only the "master" client can write to, while every other client has their own local db that are not synced. These clients simple read from the sqlite file and merge any changes to their local db. Any local changes are saved as a changesets in a json file that gets synced along with the sqlite file. The master client then merges the changeset json files to the sqlite db.

I think in this system I would also like to name the changeset files to be the node id. But then it is important for the node id to always stay the same for a device. So it is OK that I cannot set the ID, but some way to have a persistent node id for a client would be nice.

Sorry for the big text dump! Feel free to ignore it. :-)
